WooCommerce/Classes/Tools/InfiniteScroll/AsyncPaginationTracker.swift

```diff
@@ -0,0 +1,132 @@
+import Yosemite
+
+/// Async/await version of `PaginationTracker`, consider renaming `PaginationTracker` as deprecated and this class to `PaginationTracker`.
+/// Keeps track of the pagination for API syncing to support infinite scroll and pull-to-refresh.
+final class AsyncPaginationTracker {
+    typealias SyncFunction = (_ pageNumber: Int) async throws -> Bool
+
+    /// State of loading the next page in `ensureNextPageIsSynced`.
+    enum NextPageSyncState {
+        case syncing
+        case synced
+        case noNextPage
+    }
+
+    /// Default pagination settings.
+    enum Defaults {
+        static let pageFirstIndex = Store.Default.firstPageNumber
+    }
+
+    /// The index of the first page in the API. So far, both Woo and WP.com API have the first page index at 1.
+    private let pageFirstIndex: Int
+
+    /// Indexes of the pages that have been successfully synced.
+    private var pagesSynced = IndexSet()
+
+    /// Indexes of the pages being currently synced.
+    private var pagesBeingSynced = IndexSet()
+
+    /// Whether there might be more pages to fetch from the API, set by the sync function.
+    private var hasNextPage: Bool = true
+
+    /// Returns the highest page number that has been successfully synced, if any.
+    private var highestPageSynced: Int? {
+        pagesSynced.max()
+    }
+
+    /// Returns the highest page number that is currently being synced, if any.
+    private var highestPageBeingSynced: Int? {
+        pagesBeingSynced.max()
+    }
+
+    /// Designated Initializer
+    init(pageFirstIndex: Int = Defaults.pageFirstIndex) {
+        self.pageFirstIndex = pageFirstIndex
+    }
+
+    /// Should be called whenever a scroll position is approaching the end of the list for infinite scroll support.
+    /// This method will:
+    ///     1.  Proceed only if there is next page to sync.
+    ///     2.  Verify if the next page isn't currently being synced.
+    ///     3.  Proceed syncing the next page.
+    func ensureNextPageIsSynced(syncFunction: @escaping SyncFunction) async throws -> NextPageSyncState {
```

Comment:
I considered passing the sync function through init, but the sync involves states in self (`allItems` at least) so it cannot be passed in the initializer. I agree it feels repetitive passing for each method. We could consider updating this function to be static functional without any states, but that requires some refactoring and I'd prefer leaving it for later.

Reply:
Ah yeah, of course. OK, let's keep it like this for now, I don't think it'll be realistic to make it static, because then the consumer will be really limited about what it can do with the state it stores, and testing will be a pain.